metadata: Detailed Dataset Authorship Metadata

mofosyne opened the pull request:
On second thought, the dataset might actually be more important than the models. So we may want to add detailed name, author, url, etc... for it as well.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

mofosyne committed:
py: let users add full base model and dataset to model_card

mofosyne commented:
**set to draft as it's just a concept I came up with and might need a second opinion if this PR makes sense**

mofosyne commented:
Added some extra tests for extra confidence with this feature.

Also if we are doing this, we might as well go all the way and let people fully enter in the authorship metadata in base model and datasets in model card.

mofosyne commented:
@compilade if you got some time, have a look at this as I think this approach will at least be more intuitive for model card writers.

Galunid commented on examples/convert_legacy_llama.py:
```diff
@@ -849,12 +851,32 @@ def add_meta_model(self, params: Params, metadata: gguf.Metadata | None) -> None
+                for key, dataset_entry in enumerate(metadata.datasets):
+                    if "name" in dataset_entry:
+                        self.gguf.add_dataset_name(key, dataset_entry["name"])
+                    if "author" in dataset_entry:
+                        self.gguf.add_dataset_author(key, dataset_entry["author"])
+                    if "version" in dataset_entry:
+                        self.gguf.add_dataset_version(key, dataset_entry["version"])
+                    if "organization" in dataset_entry:
+                        self.gguf.add_dataset_organization(key, dataset_entry["organization"])
+                    if "description" in dataset_entry:
+                        self.gguf.add_dataset_description(key, dataset_entry["description"])
+                    if "url" in dataset_entry:
+                        self.gguf.add_dataset_url(key, dataset_entry["url"])
+                    if "doi" in dataset_entry:
+                        self.gguf.add_dataset_doi(key, dataset_entry["doi"])
+                    if "uuid" in dataset_entry:
+                        self.gguf.add_dataset_uuid(key, dataset_entry["uuid"])
+                    if "repo_url" in dataset_entry:
+                        self.gguf.add_dataset_repo_url(key, dataset_entry["repo_url"])
```
Those should be `elif` instead.

Galunid commented on gguf-py/gguf/metadata.py:
```diff
@@ -502,9 +549,29 @@ def set_gguf_meta_model(self, gguf_writer: gguf.GGUFWriter):
+            for key, dataset_entry in enumerate(self.datasets):
+                if "name" in dataset_entry:
+                    gguf_writer.add_dataset_name(key, dataset_entry["name"])
+                if "author" in dataset_entry:
+                    gguf_writer.add_dataset_author(key, dataset_entry["author"])
+                if "version" in dataset_entry:
+                    gguf_writer.add_dataset_version(key, dataset_entry["version"])
+                if "organization" in dataset_entry:
+                    gguf_writer.add_dataset_organization(key, dataset_entry["organization"])
+                if "description" in dataset_entry:
+                    gguf_writer.add_dataset_description(key, dataset_entry["description"])
+                if "url" in dataset_entry:
+                    gguf_writer.add_dataset_url(key, dataset_entry["url"])
+                if "doi" in dataset_entry:
+                    gguf_writer.add_dataset_doi(key, dataset_entry["doi"])
+                if "uuid" in dataset_entry:
+                    gguf_writer.add_dataset_uuid(key, dataset_entry["uuid"])
+                if "repo_url" in dataset_entry:
+                    gguf_writer.add_dataset_repo_url(key, dataset_entry["repo_url"])
```
Same as earlier (`elif`)

Galunid replied:
my bad

mofosyne replied:
Ah no problem. Thanks for checking anyway. Hopefully ready to merge in

compilade reviewed (commented):
Not sure if this will coexist well with how HuggingFace processes the model cards, see <https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/repocard_data.py>.

Other than that, the implementation seems sound.

It seems like the only breaking change is the removal of `Keys.General.DATASETS`, which seems reasonable.

mofosyne commented on gguf-py/gguf/metadata.py:
```diff
@@ -364,18 +366,64 @@ def use_array_model_card_metadata(metadata_key: str, model_card_key: str):
+                                dataset["version"] = version
+                            if org_component is not None and dataset_name_component is not None:
+                                dataset["repo_url"] = f"https://huggingface.co/{org_component}/{dataset_name_component}"
+                    elif isinstance(dataset_id, dict):
+                        dataset = dataset_id
```
Unsure what to do about that... but since this bit would be an 'extension' to huggingface approach... we could alert huggingface to make the change on their end and warn model writers to not use this 'feature' until huggingface accepts our proposed extension.

mofosyne replied:
Addressed by separating into separate PR

mofosyne commented:
@compilade looks like the HF team would have to do lots of refactoring to support both dicts and lists to support this concept fully. So at least according to @Wauplin it is unlikely they will be able to support more complex metadata that we are proposing.

How would you prefer to approach this? I'm still proposing the refactoring of datasets in the main GGUF standard... so we can either separate just the metadata refactoring of datasets... or we can just merge as is... but note in documentation a warning that dict is not supported in HuggingFace (Which would give HF time to support it eventually).

mofosyne committed:
py: Detailed datasets metadata in gguf kv store

mofosyne commented:
Okay, decided to refactor to move the 'dicts' in modelcard feature into a separate PR.

Kept the 'url' in string as I don't think it's as much of a leap for the HF team to adopt if they are okay with it... plus it's already in string format so won't break at least the string typecheck on their side.

mofosyne commented:
Setting to draft until I have an approach to address https://github.com/huggingface/huggingface_hub/issues/2479#issuecomment-2324841752 . Ergo their recommendation is to shift the more advance dict usage into a separate model card metadata property.

mofosyne committed:
py: Add base_model_sources and dataset_sources to metadata heuristics

This is to address "Model Card: Allow for dicts in datasets and base_model and also update spec" in https://github.com/huggingface/huggingface_hub/issues/2479 where we would like to add detailed metadata support for both base model and datashet but in a way that huggingface will eventually be able to support (They are currently using either a string or string list... we will be using a list of dict which would be extensible). They recommended creating a seperate metadata property for this.

mofosyne commented:
@compilade @Galunid after a while, I think I shall suggest that users would canonically use `base_model_sources` and  `dataset_sources` if they want to include detailed metadata for parent base model and datasets.

This is picked to avoid conflict with [Hugging Face Hub's existing definition for base_model and datasets which they said is restricted to just str and str list](https://github.com/huggingface/huggingface_hub/blob/e9cd695d7bd9e81b4eceb8f4da557a0cfa387b99/src/huggingface_hub/repocard_data.py#L249-L251). For them to more easily integrate our suggestion, they request we pick a different metadata property name.

Took some time to think it over. I think the below makes most sense.

* `base_model_sources` (`List[dict]`, *optional*)
* `dataset_sources` (`List[dict]`, *optional*)

mofosyne commented:
@ggerganov can you review this PR? It's been though a few revision and both @compilade and @lin72h gave the thumbs up at least (but I'm guessing they would like someone more experienced to put the final stamp of approval).

I've looked thought a few other PRs I've created (and closed a few), but I think this one is still relevant.

ggerganov reviewed (approved):
Seems ok to merge. I'm not familiar with datasets usage, but as long as it does not interfere with normal model usage, it should be ok.

mofosyne commented:
Appreciated. Had a quick lookover again, regarding any potential interference to normal model usage... it will read normal model metadata in huggingface in a backwards compatible way. Since we can deal with both list of string and list of dict. It's only HuggingFace's code that is a bit inflexible, hence their request to have a separate field for detailed metadata on model and dataset.
